src/c/tb_client/context.zig

```diff
@@ -13,67 +30,278 @@ pub const ContextImplementation = struct {
             context.implementation = .{
                 .submit_fn = Context.on_submit,
                 .deinit_fn = Context.on_deinit,
             };
 
-            const addresses = @ptrCast([*]const u8, addresses_ptr)[0..addresses_len];
+            log.debug("init: initializing thread.", .{});
             context.thread.init(
-                allocator,
-                cluster_id,
-                addresses,
-                num_packets,
-                Context.on_completion,
+                context,
             ) catch |err| {
-                allocator.destroy(context);
-                return switch (err) {
-                    error.Unexpected => .unexpected,
-                    error.OutOfMemory => .out_of_memory,
-                    error.InvalidAddress => .invalid_address,
-                    error.SystemResources => .system_resources,
-                    error.NetworkSubsystemFailed => .network_subsystem,
+                log.err("failed to initalize thread: {}", .{err});
+                return err;
+            };
+            errdefer context.thread.deinit(context.allocator);
+
+            return context;
+        }
+
+        pub fn destroy(self: *Context) void {
+            self.thread.deinit();
+
+            self.client.deinit(self.allocator);
+            self.message_bus.deinit(self.allocator);
+            self.message_pool.deinit(self.allocator);
+            self.io.deinit();
+
+            self.allocator.free(self.addresses);
+            self.allocator.free(self.packets);
+            self.allocator.destroy(self);
+        }
+
+        pub fn tick(self: *Context) void {
+            self.client.tick();
+        }
+
+        pub fn run(self: *Context) void {
+            while (!self.thread.signal.is_shutdown()) {
+                self.client.tick();
+                self.io.run_for_ns(config.tick_ms * std.time.ns_per_ms) catch |err| {
+                    log.err("IO.run() failed with {}", .{err});
+                    @panic("IO.run() failed.");
                 };
+            }
+        }
+
+        pub fn request(self: *Context, packet: *Packet) void {
+            const message = self.message_pool.get_message();
+            defer self.message_pool.unref(message);
+            self.available_messages -= 1;
+
+            // Get the size of each request structure in the packet.data:
+            const request_size: usize = operation_size_of(packet.operation) orelse {
+                return self.on_complete(packet, error.InvalidOperation);
             };
 
-            var list = tb_packet_list_t{};
-            for (context.thread.packets) |*packet| {
-                list.push(tb_packet_list_t.from(packet));
+            // Make sure the packet.data size is correct:
+            const readable = packet.data[0..packet.data_size];
+            if (readable.len == 0 or readable.len % request_size != 0) {
+                return self.on_complete(packet, error.InvalidDataSize);
             }
```

Comment:
This data is provided by the application, so we must validate that it is not empty and it is a multiple of the `Event` size since the application cannot split packets.

Reply:
That makes sense; thank you.